File: backend/src/domain/repository/IUserRepository.ts
import { GlobalProfile } from '../entities/GlobalProfile';
import { OrganizationMember } from '../entities/OrganizationMember';
import { AcademicRecord } from '../entities/AcademicRecord';
import { Achievement } from '../entities/Achievement';
import { SavedItem } from '../entities/SavedItem';
import { Notification } from '../entities/Notification';
import { PeerVerification } from '../entities/PeerVerification';

/**
 * Repository interface for user-related operations
 * Handles user profiles, achievements, saved items, notifications, etc.
 */
export interface IUserRepository {
  // Profile Management
  findGlobalProfileByUserId(userId: number): Promise<GlobalProfile | null>;
  updateGlobalProfile(
    userId: number,
    profileData: Partial<GlobalProfile>,
  ): Promise<GlobalProfile>;

  // Organization Membership
  joinOrganization(
    userId: number,
    organizationId: number,
    memberData: Partial<OrganizationMember>,
  ): Promise<OrganizationMember>;
  leaveOrganization(memberId: number): Promise<boolean>;
  findMemberByUserAndOrg(
    userId: number,
    organizationId: number,
  ): Promise<OrganizationMember | null>;
  findMemberById(memberId: number): Promise<OrganizationMember | null>;
  findOrganizationsByUser(userId: number): Promise<OrganizationMember[]>;
  updateMemberProfile(
    memberId: number,
    memberData: Partial<OrganizationMember>,
  ): Promise<OrganizationMember>;

  // Academic Record Management
  createAcademicRecord(
    recordData: Partial<AcademicRecord>,
  ): Promise<AcademicRecord>;
  updateAcademicRecord(
    recordId: number,
    recordData: Partial<AcademicRecord>,
  ): Promise<AcademicRecord>;
  deleteAcademicRecord(recordId: number): Promise<boolean>;
  findAcademicRecordsByMember(memberId: number): Promise<AcademicRecord[]>;
  findAcademicRecordById(recordId: number): Promise<AcademicRecord | null>;

  // Achievement Management
  createAchievement(
    achievementData: Partial<Achievement>,
  ): Promise<Achievement>;
  updateAchievement(
    achievementId: number,
    achievementData: Partial<Achievement>,
  ): Promise<Achievement>;
  deleteAchievement(achievementId: number): Promise<boolean>;
  findAchievementsByMember(
    memberId: number,
    page: number,
    limit: number,
    filters?: { status?: string },
  ): Promise<{ achievements: Achievement[]; total: number }>;
  findAchievementById(achievementId: number): Promise<Achievement | null>;
  approveAchievement(achievementId: number): Promise<Achievement>;
  rejectAchievement(achievementId: number): Promise<Achievement>;

  // Saved Items Management
  saveItem(
    memberId: number,
    itemType: string,
    itemId: number,
    note?: string,
  ): Promise<SavedItem>;
  unsaveItem(
    memberId: number,
    itemType: string,
    itemId: number,
  ): Promise<boolean>;
  findSavedItemsByMember(
    memberId: number,
    page: number,
    limit: number,
    filters?: { itemType?: string },
  ): Promise<{ savedItems: SavedItem[]; total: number }>;
  checkItemSaved(
    memberId: number,
    itemType: string,
    itemId: number,
  ): Promise<boolean>;

  // Notification Management
  createNotification(
    notificationData: Partial<Notification>,
  ): Promise<Notification>;
  findNotificationsByMember(
    memberId: number,
    page: number,
    limit: number,
    unreadOnly?: boolean,
  ): Promise<{ notifications: Notification[]; total: number }>;
  markNotificationAsRead(notificationId: number): Promise<Notification>;
  markAllNotificationsAsRead(memberId: number): Promise<boolean>;
  deleteNotification(notificationId: number): Promise<boolean>;
  countUnreadNotifications(memberId: number): Promise<number>;

  // Peer Verification Management
  createPeerVerification(
    targetMemberId: number,
    verifierMemberId: number,
  ): Promise<PeerVerification>;
  findPeerVerificationsByTarget(
    targetMemberId: number,
  ): Promise<PeerVerification[]>;
  findPeerVerificationsByVerifier(
    verifierMemberId: number,
  ): Promise<PeerVerification[]>;
  checkPeerVerificationExists(
    targetMemberId: number,
    verifierMemberId: number,
  ): Promise<boolean>;
  countPeerVerifications(targetMemberId: number): Promise<number>;

  // User Statistics
  getUserStatistics(memberId: number): Promise<{
    totalAchievements: number;
    approvedAchievements: number;
    totalSavedItems: number;
    unreadNotifications: number;
    peerVerifications: number;
  }>;
}

File: backend/src/main/java/com/service/backend/othermodule/domain/repository/IUserRepository.java
package com.service.backend.othermodule.domain.repository;

import com.service.backend.authmodule.domain.entity.GlobalProfile;
import com.service.backend.othermodule.domain.entity.OrganizationMember;
import com.service.backend.othermodule.domain.entity.AcademicRecord;
import com.service.backend.othermodule.domain.entity.Achievement;
import com.service.backend.othermodule.domain.entity.SavedItem;
import com.service.backend.othermodule.domain.entity.Notification;
import com.service.backend.othermodule.domain.entity.PeerVerification;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.Map;

/**
 * Repository interface for user-related operations
 * Handles user profiles, achievements, saved items, notifications, etc.
 */
public interface IUserRepository {

    // Profile Management
    Mono<GlobalProfile> findGlobalProfileByUserId(Long userId);
    Mono<GlobalProfile> updateGlobalProfile(Long userId, GlobalProfile profileData);

    // Organization Membership
    Mono<OrganizationMember> joinOrganization(Long userId, Long organizationId, OrganizationMember memberData);
    Mono<Boolean> leaveOrganization(Long memberId);
    Mono<OrganizationMember> findMemberByUserAndOrg(Long userId, Long organizationId);
    Mono<OrganizationMember> findMemberById(Long memberId);
    Flux<OrganizationMember> findOrganizationsByUser(Long userId);
    Mono<OrganizationMember> updateMemberProfile(Long memberId, OrganizationMember memberData);

    // Academic Record Management
    Mono<AcademicRecord> createAcademicRecord(AcademicRecord recordData);
    Mono<AcademicRecord> updateAcademicRecord(Long recordId, AcademicRecord recordData);
    Mono<Boolean> deleteAcademicRecord(Long recordId);
    Flux<AcademicRecord> findAcademicRecordsByMember(Long memberId);
    Mono<AcademicRecord> findAcademicRecordById(Long recordId);

    // Achievement Management
    Mono<Achievement> createAchievement(Achievement achievementData);
    Mono<Achievement> updateAchievement(Long achievementId, Achievement achievementData);
    Mono<Boolean> deleteAchievement(Long achievementId);
    Mono<Map<String, Object>> findAchievementsByMember(Long memberId, int page, int limit, Map<String, Object> filters);
    Mono<Achievement> findAchievementById(Long achievementId);
    Mono<Achievement> approveAchievement(Long achievementId);
    Mono<Achievement> rejectAchievement(Long achievementId);

    // Saved Items Management
    Mono<SavedItem> saveItem(Long memberId, String itemType, Long itemId, String note);
    Mono<Boolean> unsaveItem(Long memberId, String itemType, Long itemId);
    Mono<Map<String, Object>> findSavedItemsByMember(Long memberId, int page, int limit, Map<String, Object> filters);
    Mono<Boolean> checkItemSaved(Long memberId, String itemType, Long itemId);

    // Notification Management
    Mono<Notification> createNotification(Notification notificationData);
    Mono<Map<String, Object>> findNotificationsByMember(Long memberId, int page, int limit, Boolean unreadOnly);
    Mono<Notification> markNotificationAsRead(Long notificationId);
    Mono<Boolean> markAllNotificationsAsRead(Long memberId);
    Mono<Boolean> deleteNotification(Long notificationId);
    Mono<Long> countUnreadNotifications(Long memberId);

    // Peer Verification Management
    Mono<PeerVerification> createPeerVerification(Long targetMemberId, Long verifierMemberId);
    Flux<PeerVerification> findPeerVerificationsByTarget(Long targetMemberId);
    Flux<PeerVerification> findPeerVerificationsByVerifier(Long verifierMemberId);
    Mono<Boolean> checkPeerVerificationExists(Long targetMemberId, Long verifierMemberId);
    Mono<Long> countPeerVerifications(Long targetMemberId);

    // User Statistics
    Mono<Map<String, Object>> getUserStatistics(Long memberId);
}
